# Подготовка файлов и программ

## Установка HISAT2 (для выравниваия RNA-seq чтений на геном)

In [ ]:
!apt-get update

In [ ]:
!apt-get install hisat2

In [ ]:
!hisat2 --version

## Установка sra-toolkit (для скачивания .fastq файлов из NCBI)

In [ ]:
!apt-get install sra-toolkit

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  blends-common libkdf5-2 libmagic-mgc libmagic1 libmbedcrypto1 libmbedtls10
  libncbi-vdb2 libncbi-wvdb2 med-config menu
Suggested packages:
  blends-doc file menu-l10n gksu | kde-runtime | ktsuss
The following NEW packages will be installed:
  blends-common libkdf5-2 libmagic-mgc libmagic1 libmbedcrypto1 libmbedtls10
  libncbi-vdb2 libncbi-wvdb2 med-config menu sra-toolkit
0 upgraded, 11 newly installed, 0 to remove and 37 not upgraded.
Need to get 4,042 kB of archives.
After this operation, 22.6 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libmagic-mgc amd64 1:5.32-2ubuntu0.4 [184 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libmagic1 amd64 1:5.32-2ubuntu0.4 [68.6 kB]
Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 menu amd64 2.1.4

## Установка FastQC

In [22]:
!java -version

openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.18.04)
OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.18.04, mixed mode, sharing)


In [34]:
!unzip fastqc_v0.11.9.zip

Archive:  fastqc_v0.11.9.zip
  inflating: FastQC/cisd-jhdf5.jar   
   creating: FastQC/Configuration/
  inflating: FastQC/Configuration/adapter_list.txt  
  inflating: FastQC/Configuration/contaminant_list.txt  
  inflating: FastQC/Configuration/limits.txt  
  inflating: FastQC/fastqc           
  inflating: FastQC/fastqc_icon.ico  
   creating: FastQC/Help/
   creating: FastQC/Help/1 Introduction/
   creating: FastQC/Help/1 Introduction/.svn/
  inflating: FastQC/Help/1 Introduction/.svn/entries  
   creating: FastQC/Help/1 Introduction/.svn/props/
   creating: FastQC/Help/1 Introduction/.svn/text-base/
  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  
   creating: FastQC/Help/1 Introduction/.svn/tmp/
   creating: FastQC/Help/1 Introduction/.svn/tmp/props/
  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  
   creating: FastQC/Help/2 Basic Operations/
   creating: FastQC/Help/2 Basic Operations/.svn/
  inflating: FastQC/Help/2 Basic

In [35]:
!chmod a+x FastQC/fastqc

In [36]:
!./FastQC/fastqc --help


            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

	fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of a number of different modules, each one of 
    which will help to identify a different potential type of problem in your
    data.
    
    If no files to process are specified on the command line then the program
    will start as an interactive graphical application.  If files are provided
    on the command line then the program will run with no user interaction
    required.  In this mode it is suitable for inclusion into a standardised
    analysis pipeline.
    
    The options for the program as as follows:
    
    -h --help       Print this help file and exit
    
    -v --version    Print the version of the program and exit

## Установка HTSeq-count

In [18]:
!pip install HTSeq

     |████████████████████████████████| 1.5 MB 4.3 MB/s 
     |████████████████████████████████| 14.9 MB 18.3 MB/s 


In [24]:
pip install --upgrade numpy

     |████████████████████████████████| 15.7 MB 4.2 MB/s 
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.5
    Uninstalling numpy-1.19.5:
      Successfully uninstalled numpy-1.19.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
yellowbrick 1.3.post1 requires numpy<1.20,>=1.16.0, but you have numpy 1.21.4 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.
albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you have imgaug 0.2.9 which is incompatible.


In [25]:
!htseq-count --version

1.99.2


## Скачиваем RNA-seq данные

* Перепрограммированные образцы: 	SRR3414629, SRR3414630, SRR3414631
* Контрольные образцы:		SRR3414635, SRR3414636, SRR3414637

**Один образец скачивается 5-10 мин!!**

**Если не получается скачать с помощью fastq-dump, можно скачать все образцы в едином архиве по ссылке (занимает 1.5 - 2 часа):**

In [21]:
!time wget http://83.149.211.146:23194/~ivan/data/hse_teaching/rna_seq_2021.tgz

--2021-12-03 07:15:48--  http://83.149.211.146:23194/~ivan/data/hse_teaching/rna_seq_2021.tgz
Connecting to 83.149.211.146:23194... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6471183400 (6.0G) [application/x-gzip]
Saving to: ‘rna_seq_2021.tgz’

rna_seq_2021.tgz    100%[===================>]   6.03G  3.21MB/s    in 32m 38s 

2021-12-03 07:48:27 (3.15 MB/s) - ‘rna_seq_2021.tgz’ saved [6471183400/6471183400]


real	32m38.983s
user	0m11.003s
sys	1m38.874s


Распаковываем архив:

In [26]:
!time tar -xzvf rna_seq_2021.tgz

rna_seq_2021/
rna_seq_2021/_tmp.txt
rna_seq_2021/SRR3414629_1.fastq
rna_seq_2021/SRR3414636_1.fastq
rna_seq_2021/SRR3414631_1.fastq
rna_seq_2021/SRR3414635_1.fastq
rna_seq_2021/SRR3414637_1.fastq
rna_seq_2021/SRR3414630_1.fastq

real	9m49.489s
user	4m2.102s
sys	1m37.662s


## Скачиваем геном мыши mm10 (проиндексированный для HISAT2)

In [30]:
!wget https://genome-idx.s3.amazonaws.com/hisat/mm10_genome.tar.gz

--2021-12-03 08:10:37--  https://genome-idx.s3.amazonaws.com/hisat/mm10_genome.tar.gz
Resolving genome-idx.s3.amazonaws.com (genome-idx.s3.amazonaws.com)... 52.217.38.12
Connecting to genome-idx.s3.amazonaws.com (genome-idx.s3.amazonaws.com)|52.217.38.12|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3804366597 (3.5G) [binary/octet-stream]
Saving to: ‘mm10_genome.tar.gz’

mm10_genome.tar.gz  100%[===================>]   3.54G  16.3MB/s    in 4m 5s   

2021-12-03 08:14:42 (14.8 MB/s) - ‘mm10_genome.tar.gz’ saved [3804366597/3804366597]



In [31]:
!tar -xzvf mm10_genome.tar.gz

mm10/
mm10/genome.8.ht2
mm10/genome.5.ht2
mm10/make_mm10.sh
mm10/genome.7.ht2
mm10/genome.6.ht2
mm10/genome.4.ht2
mm10/genome.3.ht2
mm10/genome.1.ht2
mm10/genome.2.ht2


## Скачиваем аннотацию генов GENCODE для генома мыши mm10

In [32]:
!wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz

--2021-12-03 08:15:33--  http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz
Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.197.74
Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.197.74|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 28542432 (27M) [application/octet-stream]
Saving to: ‘gencode.vM25.annotation.gtf.gz’

gencode.vM25.annota 100%[===================>]  27.22M   260KB/s    in 2m 21s  

2021-12-03 08:17:55 (197 KB/s) - ‘gencode.vM25.annotation.gtf.gz’ saved [28542432/28542432]



In [38]:
!gzip -d /content/sample_data/gencode.vM25.annotation.gtf.gz

gzip: /content/sample_data/gencode.vM25.annotation.gtf.gz: No such file or directory


# Выравнивание чтений

**Ниже приведены примеры запуска только для одного файла. Вам следует сделать это для всех 6-ти файлов.**

* Для каждого образца запускаем FastQC. Скачиваем полученные html файлы и проверяем есть ли проблемы с файлами. Можно также установить multiqc и получить объединенный .html файл. Полученные результаты приводим в отчете на Github-е.



In [39]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414636_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414636_1.fastq -S SRR3414636_1.sam  2>  SRR3414636.hisat
!cat SRR3414636.hisat
!grep -P '^@|NH:i:1$' SRR3414636_1.sam > SRR3414636.uniq.sam
!rm -v SRR3414636_1.sam
!grep -v '^@' SRR3414636.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414636.uniq.sam  gencode.vM25.annotation.gtf > SRR3414636.counts
! grep '^__' SRR3414636.counts

Started analysis of SRR3414636_1.fastq
Approx 5% complete for SRR3414636_1.fastq
Approx 10% complete for SRR3414636_1.fastq
Approx 15% complete for SRR3414636_1.fastq
Approx 20% complete for SRR3414636_1.fastq
Approx 25% complete for SRR3414636_1.fastq
Approx 30% complete for SRR3414636_1.fastq
Approx 35% complete for SRR3414636_1.fastq
Approx 40% complete for SRR3414636_1.fastq
Approx 45% complete for SRR3414636_1.fastq
Approx 50% complete for SRR3414636_1.fastq
Approx 55% complete for SRR3414636_1.fastq
Approx 60% complete for SRR3414636_1.fastq
Approx 65% complete for SRR3414636_1.fastq
Approx 70% complete for SRR3414636_1.fastq
Approx 75% complete for SRR3414636_1.fastq
Approx 80% complete for SRR3414636_1.fastq
Approx 85% complete for SRR3414636_1.fastq
Approx 90% complete for SRR3414636_1.fastq
Approx 95% complete for SRR3414636_1.fastq
Analysis complete for SRR3414636_1.fastq

real	11m25.997s
user	19m10.610s
sys	1m12.000s
20307147 reads; of these:
  20307147 (100.00%) were unpai

In [40]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414637_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414637_1.fastq -S SRR3414637_1.sam  2>  SRR3414637.hisat
!cat SRR3414637.hisat
!grep -P '^@|NH:i:1$' SRR3414637_1.sam > SRR3414637.uniq.sam
!rm -v SRR3414637_1.sam
!grep -v '^@' SRR3414637.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414637.uniq.sam  gencode.vM25.annotation.gtf > SRR3414637.counts
! grep '^__' SRR3414637.counts

Started analysis of SRR3414637_1.fastq
Approx 5% complete for SRR3414637_1.fastq
Approx 10% complete for SRR3414637_1.fastq
Approx 15% complete for SRR3414637_1.fastq
Approx 20% complete for SRR3414637_1.fastq
Approx 25% complete for SRR3414637_1.fastq
Approx 30% complete for SRR3414637_1.fastq
Approx 35% complete for SRR3414637_1.fastq
Approx 40% complete for SRR3414637_1.fastq
Approx 45% complete for SRR3414637_1.fastq
Approx 50% complete for SRR3414637_1.fastq
Approx 55% complete for SRR3414637_1.fastq
Approx 60% complete for SRR3414637_1.fastq
Approx 65% complete for SRR3414637_1.fastq
Approx 70% complete for SRR3414637_1.fastq
Approx 75% complete for SRR3414637_1.fastq
Approx 80% complete for SRR3414637_1.fastq
Approx 85% complete for SRR3414637_1.fastq
Approx 90% complete for SRR3414637_1.fastq
Approx 95% complete for SRR3414637_1.fastq
Analysis complete for SRR3414637_1.fastq

real	11m40.763s
user	19m30.827s
sys	1m19.273s
20385570 reads; of these:
  20385570 (100.00%) were unpai

In [41]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414635_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414635_1.fastq -S SRR3414635_1.sam  2>  SRR3414635.hisat
!cat SRR3414635.hisat
!grep -P '^@|NH:i:1$' SRR3414635_1.sam > SRR3414635.uniq.sam
!rm -v SRR3414635_1.sam
!grep -v '^@' SRR3414635.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414635.uniq.sam  gencode.vM25.annotation.gtf > SRR3414635.counts
! grep '^__' SRR3414635.counts

Started analysis of SRR3414635_1.fastq
Approx 5% complete for SRR3414635_1.fastq
Approx 10% complete for SRR3414635_1.fastq
Approx 15% complete for SRR3414635_1.fastq
Approx 20% complete for SRR3414635_1.fastq
Approx 25% complete for SRR3414635_1.fastq
Approx 30% complete for SRR3414635_1.fastq
Approx 35% complete for SRR3414635_1.fastq
Approx 40% complete for SRR3414635_1.fastq
Approx 45% complete for SRR3414635_1.fastq
Approx 50% complete for SRR3414635_1.fastq
Approx 55% complete for SRR3414635_1.fastq
Approx 60% complete for SRR3414635_1.fastq
Approx 65% complete for SRR3414635_1.fastq
Approx 70% complete for SRR3414635_1.fastq
Approx 75% complete for SRR3414635_1.fastq
Approx 80% complete for SRR3414635_1.fastq
Approx 85% complete for SRR3414635_1.fastq
Approx 90% complete for SRR3414635_1.fastq
Approx 95% complete for SRR3414635_1.fastq
Analysis complete for SRR3414635_1.fastq

real	12m10.609s
user	20m25.416s
sys	1m17.310s
20956475 reads; of these:
  20956475 (100.00%) were unpai

In [42]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414631_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414631_1.fastq -S SRR3414631_1.sam  2>  SRR3414631.hisat
!cat SRR3414631.hisat
!grep -P '^@|NH:i:1$' SRR3414631_1.sam > SRR3414631.uniq.sam
!rm -v SRR3414631_1.sam
!grep -v '^@' SRR3414631.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414631.uniq.sam  gencode.vM25.annotation.gtf > SRR3414631.counts
! grep '^__' SRR3414631.counts

Started analysis of SRR3414631_1.fastq
Approx 5% complete for SRR3414631_1.fastq
Approx 10% complete for SRR3414631_1.fastq
Approx 15% complete for SRR3414631_1.fastq
Approx 20% complete for SRR3414631_1.fastq
Approx 25% complete for SRR3414631_1.fastq
Approx 30% complete for SRR3414631_1.fastq
Approx 35% complete for SRR3414631_1.fastq
Approx 40% complete for SRR3414631_1.fastq
Approx 45% complete for SRR3414631_1.fastq
Approx 50% complete for SRR3414631_1.fastq
Approx 55% complete for SRR3414631_1.fastq
Approx 60% complete for SRR3414631_1.fastq
Approx 65% complete for SRR3414631_1.fastq
Approx 70% complete for SRR3414631_1.fastq
Approx 75% complete for SRR3414631_1.fastq
Approx 80% complete for SRR3414631_1.fastq
Approx 85% complete for SRR3414631_1.fastq
Approx 90% complete for SRR3414631_1.fastq
Approx 95% complete for SRR3414631_1.fastq
Analysis complete for SRR3414631_1.fastq

real	14m18.967s
user	24m18.289s
sys	1m31.501s
24244069 reads; of these:
  24244069 (100.00%) were unpai

In [43]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414630_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414630_1.fastq -S SRR3414630_1.sam  2>  SRR3414630.hisat
!cat SRR3414630.hisat
!grep -P '^@|NH:i:1$' SRR3414630_1.sam > SRR3414630.uniq.sam
!rm -v SRR3414630_1.sam
!grep -v '^@' SRR3414630.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414630.uniq.sam  gencode.vM25.annotation.gtf > SRR3414630.counts
! grep '^__' SRR3414630.counts

Started analysis of SRR3414630_1.fastq
Approx 5% complete for SRR3414630_1.fastq
Approx 10% complete for SRR3414630_1.fastq
Approx 15% complete for SRR3414630_1.fastq
Approx 20% complete for SRR3414630_1.fastq
Approx 25% complete for SRR3414630_1.fastq
Approx 30% complete for SRR3414630_1.fastq
Approx 35% complete for SRR3414630_1.fastq
Approx 40% complete for SRR3414630_1.fastq
Approx 45% complete for SRR3414630_1.fastq
Approx 50% complete for SRR3414630_1.fastq
Approx 55% complete for SRR3414630_1.fastq
Approx 60% complete for SRR3414630_1.fastq
Approx 65% complete for SRR3414630_1.fastq
Approx 70% complete for SRR3414630_1.fastq
Approx 75% complete for SRR3414630_1.fastq
Approx 80% complete for SRR3414630_1.fastq
Approx 85% complete for SRR3414630_1.fastq
Approx 90% complete for SRR3414630_1.fastq
Approx 95% complete for SRR3414630_1.fastq
Analysis complete for SRR3414630_1.fastq

real	9m4.743s
user	15m0.365s
sys	0m54.384s
15244711 reads; of these:
  15244711 (100.00%) were unpaired

In [44]:
!./FastQC/fastqc  /content/rna_seq_2021/SRR3414629_1.fastq
!time hisat2 -p 3 -x mm10/genome -U /content/rna_seq_2021/SRR3414629_1.fastq -S SRR3414629_1.sam  2>  SRR3414629.hisat
!cat SRR3414629.hisat
!grep -P '^@|NH:i:1$' SRR3414629_1.sam > SRR3414629.uniq.sam
!rm -v SRR3414629_1.sam
!grep -v '^@' SRR3414629.uniq.sam | wc -l
!time htseq-count --format=sam --stranded=no SRR3414629.uniq.sam  gencode.vM25.annotation.gtf > SRR3414629.counts
! grep '^__' SRR3414629.counts

Started analysis of SRR3414629_1.fastq
Approx 5% complete for SRR3414629_1.fastq
Approx 10% complete for SRR3414629_1.fastq
Approx 15% complete for SRR3414629_1.fastq
Approx 20% complete for SRR3414629_1.fastq
Approx 25% complete for SRR3414629_1.fastq
Approx 30% complete for SRR3414629_1.fastq
Approx 35% complete for SRR3414629_1.fastq
Approx 40% complete for SRR3414629_1.fastq
Approx 45% complete for SRR3414629_1.fastq
Approx 50% complete for SRR3414629_1.fastq
Approx 55% complete for SRR3414629_1.fastq
Approx 60% complete for SRR3414629_1.fastq
Approx 65% complete for SRR3414629_1.fastq
Approx 70% complete for SRR3414629_1.fastq
Approx 75% complete for SRR3414629_1.fastq
Approx 80% complete for SRR3414629_1.fastq
Approx 85% complete for SRR3414629_1.fastq
Approx 90% complete for SRR3414629_1.fastq
Approx 95% complete for SRR3414629_1.fastq
Analysis complete for SRR3414629_1.fastq

real	12m32.643s
user	20m57.513s
sys	1m22.939s
21106089 reads; of these:
  21106089 (100.00%) were unpai

In [45]:
!pip install multiqc
!mkdir multiqc
!multiqc -o multiqc /content/rna_seq_2021

     |████████████████████████████████| 1.2 MB 4.2 MB/s 
     |████████████████████████████████| 130 kB 38.0 MB/s 
     |████████████████████████████████| 214 kB 44.2 MB/s 
     |████████████████████████████████| 46 kB 3.1 MB/s 
     |████████████████████████████████| 596 kB 44.5 MB/s 
     |████████████████████████████████| 51 kB 6.0 MB/s 
     |████████████████████████████████| 86 kB 5.3 MB/s 
  Created wheel for spectra: filename=spectra-0.0.11-py3-none-any.whl size=17488 sha256=edeb58aa9ef07ae5ded1b219b25339fff7e3d90f0245f155217106ed27f113bb
  Stored in directory: /root/.cache/pip/wheels/e3/c8/a4/5a566fafff827fcb5741589e42002686bb29990fbc7b262b5e
  Created wheel for colormath: filename=colormath-3.0.0-py3-none-any.whl size=39408 sha256=80c4b036bf0df5c86115a74bac95661d0ff1f4abf0515eea04c2c4bdb29c7077
  Stored in directory: /root/.cache/pip/wheels/1a/d7/a7/458a4632ccc8bfe436f5e9db9fef1aa84e5d243a1b4b4c7b59
  Created wheel for lzstring: filename=lzstring-1.0.4-py2.py3-none-any.whl siz

In [46]:
SRR3414629_sum = 18375888 - 1604107 - 722172
SRR3414630_sum = 13186139 - 1240295 - 480520
SRR3414631_sum = 20928945 - 1700354 - 819740
SRR3414635_sum = 18428317 - 1392186 - 760134
SRR3414636_sum = 17825380 - 1332692 - 735108
SRR3414637_sum = 17844858 - 1397650 - 710230
print('общее число чтений для SRR3414629, соответствующих хотя бы одному гену =', SRR3414629_sum)
print('общее число чтений для SRR3414630, соответствующих хотя бы одному гену =', SRR3414630_sum)
print('общее число чтений для SRR3414631, соответствующих хотя бы одному гену =', SRR3414631_sum)
print('общее число чтений для SRR3414635, соответствующих хотя бы одному гену =', SRR3414635_sum)
print('общее число чтений для SRR3414636, соответствующих хотя бы одному гену =', SRR3414636_sum)
print('общее число чтений для SRR3414637, соответствующих хотя бы одному гену =', SRR3414637_sum)

общее число чтений для SRR3414629, соответствующих хотя бы одному гену = 16049609
общее число чтений для SRR3414630, соответствующих хотя бы одному гену = 11465324
общее число чтений для SRR3414631, соответствующих хотя бы одному гену = 18408851
общее число чтений для SRR3414635, соответствующих хотя бы одному гену = 16275997
общее число чтений для SRR3414636, соответствующих хотя бы одному гену = 15757580
общее число чтений для SRR3414637, соответствующих хотя бы одному гену = 15736978


In [47]:
import pandas as pd
c1 = pd.read_csv('SRR3414635.counts', sep="\t",  names=['geneID', 'c1'])
c2 = pd.read_csv('SRR3414636.counts', sep="\t",  names=['geneID', 'c2'])
c3 = pd.read_csv('SRR3414637.counts', sep="\t",  names=['geneID', 'c3'])
r1 = pd.read_csv('SRR3414629.counts', sep="\t",  names=['geneID', 'r1'])
r2 = pd.read_csv('SRR3414630.counts', sep="\t",  names=['geneID', 'r2'])
r3 = pd.read_csv('SRR3414631.counts', sep="\t",  names=['geneID', 'r3'])
c1 = c1.merge(c2, how='left', on='geneID')
c1 = c1.merge(c3, how='left', on='geneID')
c1 = c1.merge(r1, how='left', on='geneID')
c1 = c1.merge(r2, how='left', on='geneID')
c1 = c1.merge(r3, how='left', on='geneID')
all_counts = c1[:-5]
all_counts

geneID     c1     c2     c3   r1     r2     r3
0       ENSMUSG00000000001.4   3431   3504   4031  NaN   3919   5700
1      ENSMUSG00000000003.15      0      0      0  NaN      0      0
2      ENSMUSG00000000028.15    150    136    152  NaN    273    468
3      ENSMUSG00000000031.16  55526  48225  56064  NaN  33249  64992
4      ENSMUSG00000000037.17     41     44     52  NaN     68     87
...                      ...    ...    ...    ...  ...    ...    ...
55396   ENSMUSG00000118655.1      0      0      0  NaN      1      1
55397   ENSMUSG00000118656.1      0      0      0  NaN      0      0
55398   ENSMUSG00000118657.1      0      1      0  NaN      0      4
55399   ENSMUSG00000118658.1      0      0      0  NaN      0      0
55400   ENSMUSG00000118659.1      0      0      0  NaN      0      0

[55401 rows x 7 columns]

In [48]:
from google.colab import files
all_counts.to_csv('all_counts.csv')
files.download('all_counts.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [49]:
mergedfile = open('ALL.counts', 'w')
mergedfile.write(all_counts.to_csv(sep='\t', index=False))
mergedfile.close()

In [50]:
data = {'id':['SRR3414635', 'SRR3414636', 'SRR3414637', 'SRR3414629', 'SRR3414630', 'SRR3414631'], 'condition':['control', 'control', 'control', 'reprogramming', 'reprogramming', 'reprogramming']}
all_info = pd.DataFrame(data, index =['c1', 'c2', 'c3', 'r1', 'r2', 'r3'])
all_info 

id      condition
c1  SRR3414635        control
c2  SRR3414636        control
c3  SRR3414637        control
r1  SRR3414629  reprogramming
r2  SRR3414630  reprogramming
r3  SRR3414631  reprogramming

In [51]:
infofile = open('ALL.info', 'w')
infofile.write(all_info.to_csv(sep='\t'))
infofile.close()

In [52]:
all_info.to_csv('all_info.csv')
files.download('all_info.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

* Смотрим сколько чтений не удалось приписать ни одному гену:
* `__no_feature 1332692` – столько чтений соответствует участкам генома, где не аннотировано ни одного экзона
* `__ambiguous 735108` – столько чтений могут принадлежать разным генам
* Более подробно про такие случаи можно посмотреть в описании HTSeq-count - https://htseq.readthedocs.io/en/release_0.11.1/count.html
* Итого, общее число чтений, соответствующих хотя бы одному гену равно:
17825380 – 1332692 – 735108 = 15757580
* Для каждого образца приводим эту статистику в отчете на Github


* Объединям все файлы .counts по генам в один общий файл `ALL.counts`.
* В качестве заголовка для каждого образца указывать не его SRR id, а c1, c2, c3 для контрольных образцов и r1, r2, r3 для перепрограммированных образцов.
* Загружаем полученный ALL.counts файл в Github репозиторий